### ISSUE TITLE
transient-infix-value fails for multi-valued transient-option's

### ISSUE THREAD

**Issue by jdtsmith** [magit/transient#94]
Consider the following infix:

```elisp
(define-infix-argument this-infix ()
  :key "p"
  :class transient-option
  :argument "pubs="
  :format " %k %v"
  :multi-value t
  :choices '("one" "two" "three" "four"))
```

This produces an object value which is a list, causing `transient-infix-value` to fail in its attempt concatenate the argument and value.  A simple fix:

```lisp
(cl-defmethod transient-infix-value ((obj transient-option))
  "Return (concat ARGUMENT VALUE) or nil.

ARGUMENT and VALUE are the values of the respective slots of OBJ.
If VALUE is nil, then return nil.  VALUE may be the empty string,
which is not the same as nil."
  (when-let ((value (oref obj value)))
    (if (listp value) (setq value (string-join value ",")))
    (concat (oref obj argument) value)))
```

More generally, being able to access the infix _objects_ and their set values directly yourself in the suffix command would be far preferable, rather than unpacking a lossy string (i.e. imagine if the choices had commas inside them).

**Comment by tarsius** [comment]
> This produces an object value which is a list, ...
> More generally, ...  would be far preferable ...

This will require quite a bit of thought. I recently noticed that myself when I tried to add some transients. I'll work on it but this has to be done carefully and I don't feel like doing it right now.

Having taken a quick look I don't think I did not mean for this to just work at this point. With one exception all the "multi value" infixes in magit use classes defined in `magit-transient.el` (and their methods). The exception uses `transient-file`, which also has some methods all to itself.

For now you will have to define your own derived class if you need behavior that isn't covered by the existing classes. If the method above is all that is needed: great! Define that for your class.

**Comment by jdtsmith** [comment]
OK thanks.  For now I'm abandoning `transient-infix-value` and looping over `transient-current-suffixes` looking for objects which are or inherit from the `transient-infix` class.  Then I can accumulate their values.  

I also bumped into the "hidden objects" limitation when using an `:if` predicate to attempt to toggle display of a given infix.  For example, suppose you have an infix `--derived` that is meaningless and should be hidden, unless another infix `--main` is also set.  I thought in `--derived`'s `:if` predicate I could just ask for the current value (if any) of the `--main` infix object, and hide/show accordingly.  But there is no convenient access to the _objects themselves_ I could find (other than trolling through `transient-current-suffixes`).  And in any case, `:if` and its friends apparently do not update on every refresh, but only when the transient if first displayed. So I gave that up. 

Thanks for these tools.  Slowly wrapping my head around the `transient` approach (which I'm re-purposing for interfacing with a web search API).
